src/close.rs

```
@@ -0,0 +1,99 @@
+}
+
+#[derive(Debug)]
+pub struct Message1 {
+    close_transaction: CloseTransaction,
+    sig_close_transaction: Signature,
+}
+
+impl State0 {
+    pub fn compose(&self, final_address: Address) -> Message0 {
+        Message0 { final_address }
+    }
+
+    pub fn interpret(
+        self,
+        final_address_self: Address,
+        Message0 {
+            final_address: final_address_other,
+        }: Message0,
+    ) -> State1 {
+        State1 {
+            channel: self.0,
+            final_address_self,
+            final_address_other,
+        }
+    }
+}
+
+impl State1 {
+    pub fn compose(&self) -> anyhow::Result<Message1> {
+        let (amount_a, ownership_a) = self.channel.current_state.signed_TX_s.outputs().a;
+        let (amount_b, ownership_b) = self.channel.current_state.signed_TX_s.outputs().b;
+
+        let (output_a, output_b) = if ownership_a == self.channel.x_self.public() {
+            (
+                (amount_a, self.final_address_self.clone()),
+                (amount_b, self.final_address_other.clone()),
+            )
+        } else if ownership_b == self.channel.x_self.public() {
+            (
+                (amount_a, self.final_address_other.clone()),
+                (amount_b, self.final_address_self.clone()),
+            )
+        } else {
+            anyhow::bail!("No valid output found")
+        };
+
+        let close_transaction = CloseTransaction::new(&self.channel.TX_f_body, output_a, output_b);
+
+        let sig_close_transaction = close_transaction.sign_once(self.channel.x_self.clone());
+
+        Ok(Message1 {
+            close_transaction,
+            sig_close_transaction,
+        })
+    }
+
+    pub fn interpret(
+        self,
+        Message1 {
+            close_transaction,
+            sig_close_transaction: sig_close_transaction_other,
+        }: Message1,
+    ) -> anyhow::Result<FinalState> {
+        // in a real application we would double check the amounts
```

:wrench: We could verify the signature provided by the other party. It's not critical because if we don't and it's invalid the transaction will simply fail when being broadcast. But it's nice to fail early, so that the higher level protocol can then proceed with a force close, for example.